### PR TITLE
Support username casing change in Admin UI

### PR DESCRIPTION
**Pull request by RiadGahlouz**
Summary of the changes (in less than 80 characters):
This PR adds support for username casing change in the Admin UI. There are no UI changes, simply controller flow changes.

Addresses https://github.com/NuGet/Engineering/issues/5057

### CODE REVIEW

**Review by lyndaidaii — APPROVED** [review]

**Review by mariaghiondea — APPROVED** [review]

**Review by erdembayar — DISMISSED** [review]

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 1]
```diff
@@ -123,19 +123,24 @@ public async Task<ActionResult> ChangeUsername(string oldUsername, string newUse
...
-                IsDeleted = true,
-                CreatedUtc = _dateTimeProvider.UtcNow
-            };
-
-            account.Username = newUsername;
 
             await _auditingService.SaveAuditRecordAsync(new UserAuditRecord(account, AuditedUserAction.ChangeUsername));
 
-            _userRepository.InsertOnCommit(newAccountForOldUsername);
+            if (account.Username.Equals(newUsername, StringComparison.OrdinalIgnoreCase) == false)
```
[the remark this answers is not in the document]
Comment (RiadGahlouz) [reply src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:129]: You usually don't want to use InvariantCulture in this case as for example this would be a true statement `string.Equals("ss", "ß", StringComparison.InvariantCultureIgnoreCase)` but we don't want that

**Review by advay26 — DISMISSED** [review]

**Comment by advay26** [comment]
Should this be the linked issue? https://github.com/NuGet/Engineering/issues/5057

**Review by mariaghiondea — DISMISSED** [review]

**Review by dannyjdev — COMMENTED** [review]
Apart from updating the record, there is a previous step we perform when changing username casing. We check if the user have available packages (check the playbook for more context). I think this could be added by adding a column on the first form (**Verify Account**) to include if the user has available packages. And also, please update the playbook with this new changes.

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 2]
```diff
@@ -123,32 +123,37 @@ public async Task<ActionResult> ChangeUsername(string oldUsername, string newUse
...
-            var result = new ValidateUsernameResult();
-            result.IsFormatValid = UsernameValidationRegex.IsMatch(username);
-            result.IsAvailable = _userService.FindByUsername(username, includeDeleted: true) == null;
+            var foundUser = _userService.FindByUsername(username, includeDeleted: true);
 
-            return result;
+            return new ValidateUsernameResult()
+            {
+                IsFormatValid = UsernameValidationRegex.IsMatch(username),
+                IsAvailable = foundUser == null || foundUser.Username != username // The username check is in the event where we found a user in the DB but we're doing a cAsIng change
```
Comment (zhhyu) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:155]: >foundUser == null || foundUser.Username != username [](http://example.com/codeflow?start=30&length=51)

Maybe I miss something here. For example, if the user "A" wants to change the username to the name of another existing user "B", "foundUser" (that is "B") is not null and also "foundUser.Username" ("B") is not equal to "A", then "B" is available for "A" to use?

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 3]
```diff
@@ -123,32 +123,37 @@ public async Task<ActionResult> ChangeUsername(string oldUsername, string newUse
                 return Json(HttpStatusCode.BadRequest, "New username validation failed.", JsonRequestBehavior.AllowGet);
             }
 
-            var newAccountForOldUsername = new User()
+            if (account.Username.Equals(newUsername, StringComparison.OrdinalIgnoreCase) == false)
```
Comment (zhhyu) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:126]: Is there any chance to add some test coverage? There are some corner cases that we'd like to cover for this flow.

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs (hunk 2 above)
Comment (zhhyu) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:155]: Maybe I miss something here. For example, if user "A" wants to change the name to "B" but there is another existing user named "b". As string comparisons in SQL server are case insensitive by default, "foundUser" is not null and also "foundUser.Username" ("b") is not equal to "B", "A" will be renamed to "B". However, this is the case that "nuget.org" doesn't want to allow.

Will it be better that we check whether "foundUser" belongs to the same identity?

Comment (RiadGahlouz) [reply src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs]: Added check for this as well as a unit test!

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs (hunk 3 above)
Comment (RiadGahlouz) [reply src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:144]: Done :)

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 4]
```diff
@@ -83,14 +87,28 @@ public ActionResult VerifyAccount(string accountEmailOrUsername)
...
             {
                 return Json(HttpStatusCode.BadRequest, "Username cannot be null or empty.", JsonRequestBehavior.AllowGet);
             }
 
-            var result = ValidateUsername(newUsername);
+            var result = CheckUserExist(newUsername);
+
+            if (checkOwnedPackages)
+            {
+                var oldAccount = _userService.FindByUsername(oldUsername);
```
Comment (zhhyu) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:101]: This has been checked at "CheckUserExist()". Will it be better that we exist earlier?

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 5]
```diff
@@ -83,14 +87,28 @@ public ActionResult VerifyAccount(string accountEmailOrUsername)
         }
 
         [HttpGet]
-        public ActionResult ValidateNewUsername(string newUsername)
+        public ActionResult ValidateNewUsername(string newUsername, bool checkOwnedPackages, string oldUsername = "")
```
Comment (erdembayar) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:90]: I don't see where we're passing `true` for `checkOwnedPackages` other than unit test. How we're passing value here?

Comment (erdembayar) [reply src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:90]: @RiadGahlouz 
Have you looked into my question? After this I can approve this PR.

File: src/NuGetGallery/Areas/Admin/Views/ChangeUsername/Index.cshtml [hunk 6]
```diff
@@ -57,7 +57,10 @@
 
         <i>Validate if the new username is available and has a valid format.</i><br />
 
+        <input type="text" placeholder="Old Username" autocomplete="off" autofocus style="width: 200px;" rows="1" data-bind="value: oldUsername" />
         <input type="text" placeholder="New username" autocomplete="off" autofocus style="width: 200px;" rows="1" data-bind="value: newUsername" />
+        <input type="checkbox" id="checkNonDeletedOwnedPackages" data-bind="checked: checkOwnedPackages" />
+        <label for="checkNonDeletedOwnedPackages">Check Non-Deleted Owned Packages</label>
```
Comment (agr) [inline src/NuGetGallery/Areas/Admin/Views/ChangeUsername/Index.cshtml:63]: If setting this checkbox just adds a list of owned packages, how about changing it to "List ..."?

Comment (agr) [reply src/NuGetGallery/Areas/Admin/Views/ChangeUsername/Index.cshtml:63]: Maybe even enable it automatically if new/old only differs by the case?

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 7]
```diff
@@ -116,39 +134,44 @@ public async Task<ActionResult> ChangeUsername(string oldUsername, string newUse
...
-            var result = new ValidateUsernameResult();
-            result.IsFormatValid = UsernameValidationRegex.IsMatch(username);
-            result.IsAvailable = _userService.FindByUsername(username, includeDeleted: true) == null;
+            var foundUser = _userService.FindByUsername(username, includeDeleted: true);
 
-            return result;
+            return new ValidateUsernameResult()
+            {
+                IsFormatValid = UsernameValidationRegex.IsMatch(username),
+                IsAvailable = foundUser == null || (requestor.Key == foundUser.Key && foundUser.Username != username) // The username check is in the event where we found a user in the DB but we're doing a cAsIng change
```
Comment (lyndaidaii) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:173]:
```suggestion
                IsAvailable = foundUser == null || (requestor.Key == foundUser.Key && foundUser.Username != username) // The username check is in the event where we found a user in the DB but we're doing a casing change
```

File: tests/NuGetGallery.Facts/Areas/Admin/Controllers/ChangeUsernameControllerFacts.cs [hunk 8]
```diff
@@ -211,6 +236,32 @@ public async void WhenValidAccountAndNewUsernameReturnsOkStatusCode()
                 Assert.Equal(((int)HttpStatusCode.OK), ChangeUsernameController.Response.StatusCode);
                 Assert.Equal("Account renamed successfully!", result.Data);
             }
+
+            [Fact]
+            public async void WhenValidAccountCasingChangeAndNewUsernameReturnsOkStatusCode()
+            {
+                var oldUsername = IndividualAccount.Username;
+
+                var result = await ChangeUsernameController.ChangeUsername(IndividualAccount.Username, IndividualAccount.Username.ToUpper()) as JsonResult;
```
Comment (zhhyu) [inline tests/NuGetGallery.Facts/Areas/Admin/Controllers/ChangeUsernameControllerFacts.cs:245]: >IndividualAccount.Username [](http://example.com/codeflow?start=75&length=26)

nit: "IndividualAccount.Username" -> "oldUsername"

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs (hunk 5 above)
Comment (zhhyu) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:90]: > = "" [](http://example.com/codeflow?start=111&length=5)

I am thinking whether this default value is needed, if this "oldUsername" will be passed to query DB anyway. Should the caller always pass an "oldUsername"? The API shape is changed here, so I guess that the PR updates all the callers who use this API.

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 9]
```diff
@@ -83,14 +87,28 @@ public ActionResult VerifyAccount(string accountEmailOrUsername)
...
-        public ActionResult ValidateNewUsername(string newUsername)
+        public ActionResult ValidateNewUsername(string newUsername, bool checkOwnedPackages, string oldUsername = "")
         {
             if (string.IsNullOrEmpty(newUsername))
             {
                 return Json(HttpStatusCode.BadRequest, "Username cannot be null or empty.", JsonRequestBehavior.AllowGet);
             }
 
-            var result = ValidateUsername(newUsername);
+            var oldAccount = _userService.FindByUsername(oldUsername);
```
Comment (zhhyu) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:97]: >oldAccount [](http://example.com/codeflow?start=16&length=10)

nit: oldAccount -> account

It may be better to use the same name for local variables under the same class. For example, line 130 uses "account". We can follow the same name convention and also the account itself is still the same but the name will be changed.

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs [hunk 10]
```diff
@@ -116,39 +134,44 @@ public async Task<ActionResult> ChangeUsername(string oldUsername, string newUse
                 return Json(HttpStatusCode.NotFound, "Old username account was not found.", JsonRequestBehavior.AllowGet);
             }
 
-            var newUsernameValidation = ValidateUsername(newUsername);
+            var newUsernameValidation = ValidateUsernameChange(account, newUsername);
```
Comment (zhhyu) [inline src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:137]: >newUsernameValidation [](http://example.com/codeflow?start=16&length=21)

nit: newUsernameValidation -> validatedResult

The same at line 103.

File: src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs (hunk 5 above)
Comment (agr) [reply src/NuGetGallery/Areas/Admin/Controllers/ChangeUsernameController.cs:90]: It's coming from a form input:

https://github.com/NuGet/NuGetGallery/blob/44b0ccf4375fc36d192b92e04d5328f89a3111b0/src/NuGetGallery/Areas/Admin/Views/ChangeUsername/Index.cshtml#L62

**Comment by erdembayar** [comment]
@RiadGahlouz 
I changed to draft mode since there has been no update in the past few months. Feel free to switch to `Ready to review` mode when the PR is ready.

**Comment by erdembayar** [comment]
Let's close it for now, we can re-open later.
